### PR TITLE
Connect a wallet to the api and add balance and transfer function

### PR DESCRIPTION
**Pull request by marc-aurele-besner**
## Connect a wallet to the api and add balance and transfer function

- Attach a wallet from extension, mnemonic or seed
- Query balance
- Execute a transfer
- Test all of the above

### CODE REVIEW

**Review by abhi3700 — COMMENTED** [review]
Some are minor out of all.

File: packages/auto-consensus/__test__/balances.test.ts [hunk 1]
```diff
@@ -1,21 +1,140 @@
...
+    test('Check totalIssuance return a number greater than zero', async () => {
+      // totalIssuance is an async function that returns a hex number as a string
+      const rawIssuance = await totalIssuance()
+      // Convert the hex number to a BigInt
+      const issuance = BigInt(rawIssuance.toString())
+      // Check if the issuance is greater than zero
+      expect(issuance).toBeGreaterThan(BigInt(0))
+    })
+  })
+
```
[the remark this answers is not in the document]
Comment (marc-aurele-besner) [reply packages/auto-consensus/__test__/balances.test.ts:46]: Not sure I understand, I'm not planning on implementing a function to issue token...

File: packages/auto-consensus/src/balances.ts [hunk 2]
```diff
@@ -1,4 +1,18 @@
...
+type RawBalanceData = {
+  free: any
+  reserved: any
+  frozen: any
+  flags: any
+}
+type BalanceData = {
+  free: any
+  reserved: any
+  frozen: any
```
[the remark this answers is not in the document]
Comment (marc-aurele-besner) [reply packages/auto-consensus/src/balances.ts:14]: You mean the fact that I removed flags from the returned type?
I don't understand what flags correspond to, do you know it's purpose?

[the remark this answers is not in the document]
Comment (abhi3700) [reply packages/auto-consensus/src/balances.ts:14]: I  meant shouldn't we have defined like this:
```suggestion
  free: BigInt
  reserved: BigInt
  frozen: BigInt
  flags: boolean
}
type BalanceData = {
  free: BigInt
  reserved: BigInt
  frozen: BigInt
```

Please check the type accordingly, may be it's different than `BigInt`. But you get the point.

File: packages/auto-consensus/__test__/balances.test.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (abhi3700) [reply packages/auto-consensus/__test__/balances.test.ts:46]: Ok then, it's fine.

File: packages/auto-utils/__test__/wallet.test.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+      const pair = await setupWallet({ mnemonic: TEST_MNEMONIC })
+      expect(pair.address).toEqual(TEST_ADDRESS)
+    })
+
+    test('Check setupWallet return a pair with matching private key when provided with Alice seed', async () => {
+      const pair = await setupWallet({ uri: ALICE_URI })
+      expect(pair.address).toEqual(ALICE_ADDRESS)
+    })
+
+    test('Check setupWallet return a pair with matching private key when provided with Bob seed', async () => {
```
[the remark this answers is not in the document]
Comment (marc-aurele-besner) [reply packages/auto-utils/__test__/wallet.test.ts:35]: We already check that Bob uri return Bob address, I don't see the point of writing a specific test that check if Bob does not return Alice. We are just wrapping polkadot js to make it easier to use for our ecosystem use case, I don't believe we need to re-do all the test that polkadot js is likely doing already...

File: packages/auto-consensus/__test__/balances.test.ts [hunk 4]
```diff
@@ -1,21 +1,140 @@
...
+    //   const sender = accounts[0]
+    //   const tx = await transfer(api, ALICE_ADDRESS, 1)
+
+    //   expect(() => tx.signAndSend(sender)).toThrow(
+    //     'Unreachable code should not be executed (evaluating',
+    //     // To-Do: Confirm this is the expected error message
+    //   )
+    // })
   })
```
[the remark this answers is not in the document]
Comment (marc-aurele-besner) [reply packages/auto-consensus/__test__/balances.test.ts]: This feel like trying to test the underlying polkadot js or the protocol more than this sdk

File: packages/auto-utils/__test__/wallet.test.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+  const ALICE_URI = '//Alice'
+  const ALICE_ADDRESS = '5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY'
+  const BOB_URI = '//Bob'
+  const BOB_ADDRESS = '5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty'
+
+  describe('Test setupWallet()', () => {
+    test('Check setupWallet return a pair with matching address and public key when provided with a mnemonic', async () => {
+      const pair = await setupWallet({ mnemonic: TEST_MNEMONIC })
+      expect(pair.address).toEqual(TEST_ADDRESS)
+    })
```
[the remark this answers is not in the document]
Comment (marc-aurele-besner) [reply packages/auto-utils/__test__/wallet.test.ts:28]: There is a address() function in auto-consensus that return the st encoded address, do we use the st address in any domains else than on consensus?

File: packages/auto-consensus/__test__/balances.test.ts [hunk 6]
```diff
@@ -1,21 +1,140 @@
...
+        expect(txHash).toBeDefined()
+
+        const _balanceSenderEnd = await balance(api, address(sender.address))
+        const _balanceReceiverEnd = await balance(api, address(BOB_ADDRESS))
+        expect(_balanceSenderEnd.free).toBeLessThan(_balanceSenderStart.free)
+        expect(_balanceReceiverEnd.free).toBeGreaterThan(_balanceReceiverStart.free)
+      })
+    }
+
+    // To-Do: Fix this test
```
[the remark this answers is not in the document]
Comment (marc-aurele-besner) [reply packages/auto-consensus/__test__/balances.test.ts]: I was having issue reproducing this test, and it's just give me a timeout error so I remove it

File: packages/auto-consensus/__test__/info.test.ts [hunk 7]
```diff
@@ -6,9 +6,9 @@ describe('Verify info functions', () => {
     await activate()
   })
 
-  afterAll(async () => {
-    await disconnect()
-  })
+  // afterAll(async () => {
+  //   await disconnect()
+  // })
```
[the remark this answers is not in the document]
Comment (marc-aurele-besner) [reply packages/auto-consensus/__test__/info.test.ts]: This was a mistake, thank you, it's back

File: packages/auto-utils/__test__/wallet.test.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (abhi3700) [reply packages/auto-utils/__test__/wallet.test.ts:35]: ok, then.

File: packages/auto-utils/__test__/wallet.test.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (abhi3700) [reply packages/auto-utils/__test__/wallet.test.ts:28]: >  do we use the st address in any domains else than on consensus?

As of now, we don't I guess. But may be in future.
`st`-address last time I saw it being asked in Space Acres.

**Review by abhi3700 — APPROVED** [review]
LGTM!

**Review by jfrank-summit — APPROVED** [review]

**Pull request merged**